### PR TITLE
mgr/volumes: uid, gid for subvolume create and subvolumegroup create commands

### PR DESCRIPTION
**Pull request by joscollin**
Allow setting uid, gid for subvolume create and subvolumegroup create commands

Fixes: https://tracker.ceph.com/issues/40959

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by joscollin** [comment]
@ajarr Could you please check if `ceph.dir.layout.uid` is correct?

File: src/pybind/mgr/volumes/fs/subvolume.py [hunk 1]
```diff
@@ -99,6 +103,19 @@ def create_subvolume(self, spec, size=None, namespace_isolated=True, mode=0o755,
                 xattr_val = self._get_ancestor_xattr(subvolpath, "ceph.dir.layout.pool")
             # TODO: handle error...
             self.fs.setxattr(subvolpath, xattr_key, xattr_val.encode('utf-8'), 0)
+
+            if uid is not None:
+                try:
+                    self.fs.setxattr(subvolpath, 'ceph.dir.layout.uid', uid.encode('utf-8'), 0)
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/subvolume.py]: Um, not sure where "ceph.dir.layout.uid" comes from. What I was expecting was `self.fs.chown(...)`?

Comment (joscollin) [reply src/pybind/mgr/volumes/fs/subvolume.py]: I will introduce that in cephfs.pyx and then call it.

File: src/pybind/mgr/volumes/fs/subvolume.py [hunk 2]
```diff
@@ -101,6 +105,22 @@ def create_subvolume(self, spec, size=None, namespace_isolated=True, mode=0o755,
                 xattr_val = self._get_ancestor_xattr(subvolpath, "ceph.dir.layout.pool")
             # TODO: handle error...
             self.fs.setxattr(subvolpath, xattr_key, xattr_val.encode('utf-8'), 0)
+
+            if uid is None:
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/subvolume.py]: What about usingn the uid/gid of the subvoluem group by default?

Comment (joscollin) [reply src/pybind/mgr/volumes/fs/subvolume.py]: @batrick Fixed for subvolume inherits it's group's uid and gid. 
What do you suggest for subvolume group? I'm leaving the subvolume group defaults as `0`, as `/volumes` is a prefix. But then it is also a directory.

Comment (batrick) [reply src/pybind/mgr/volumes/fs/subvolume.py]: Default 0 is probably necessary because CSI and probably Manila rely on that. @ajarr ?

**Comment by ajarr** [comment]
> Please run this through teuthology and post the results.

@joscollin I will take this up

**Comment by joscollin** [comment]
@ajarr Could you please reply to this before your teuthology run?
https://github.com/ceph/ceph/pull/30336#discussion_r342394877

File: src/pybind/mgr/volumes/fs/subvolume.py (hunk 2 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/subvolume.py]: The libcephfs client create the directories with uid and gid 0 for CSI, https://github.com/ceph/ceph/commit/033d9e743a9#diff-cfd3b6f517caccc18f7f066395e8a4bdR73 , and, likewise in libcephfs client in ceph_volume_client library for manila.

Comment (joscollin) [reply src/pybind/mgr/volumes/fs/subvolume.py]: Thanks @ajarr. 
I understand the current setting is good for mgr/volumes:
1. Subvolume inherits it's group's uid and gid by default.
2. Subvolume Group's uid and gid defaults to 0.

If these are okay, please proceed with your teuthology run. Thanks.

File: qa/tasks/cephfs/test_volumes.py [hunk 3]
```diff
@@ -396,6 +396,28 @@ def test_subvolume_group_create_with_desired_mode(self):
         self._fs_cmd("subvolumegroup", "rm", self.volname, group1)
         self._fs_cmd("subvolumegroup", "rm", self.volname, group2)
 
+    def test_subvolume_group_create_with_desired_uid_gid(self):
+        """
+        That the subvolume group can be created with the desired uid and gid and it's uid and gid matches the
```
Comment (ajarr) [inline qa/tasks/cephfs/test_volumes.py]: s/it's/its/

File: qa/tasks/cephfs/test_volumes.py [hunk 4]
```diff
@@ -432,6 +454,28 @@ def test_subvolume_create_with_desired_mode_in_group(self):
         self._fs_cmd("subvolume", "rm", self.volname, subvol3, group)
         self._fs_cmd("subvolumegroup", "rm", self.volname, group)
 
+    def test_subvolume_create_with_desired_uid_gid(self):
+        """
+        That the subvolume can be created with the desired uid and gid and it's uid and gid matches the
```
Comment (ajarr) [inline qa/tasks/cephfs/test_volumes.py]: s/it's/its/

File: qa/tasks/cephfs/test_volumes.py [hunk 5]
```diff
@@ -396,6 +396,28 @@ def test_subvolume_group_create_with_desired_mode(self):
...
+
+        # make sure it exists
+        subvolgrouppath = self._get_subvolume_group_path(self.volname, subvolgroupname)
+        self.assertNotEqual(subvolgrouppath, None)
+
+        # verify the uid and gid
+        suid = int(self.mount_a.run_shell(['stat', '-c' '%u', subvolgrouppath]).stdout.getvalue().strip())
+        sgid = int(self.mount_a.run_shell(['stat', '-c' '%g', subvolgrouppath]).stdout.getvalue().strip())
+        self.assertEqual(uid, suid)
+        self.assertEqual(gid, sgid)
```
Comment (ajarr) [inline qa/tasks/cephfs/test_volumes.py]: In other test, we clean up the subvolume groups created

File: qa/tasks/cephfs/test_volumes.py [hunk 6]
```diff
@@ -432,6 +454,28 @@ def test_subvolume_create_with_desired_mode_in_group(self):
...
+        # make sure it exists
+        subvolpath = self._get_subvolume_path(self.volname, subvolname)
+        self.assertNotEqual(subvolpath, None)
+
+        # verify the uid and gid
+        suid = int(self.mount_a.run_shell(['stat', '-c' '%u', subvolpath]).stdout.getvalue().strip())
+        sgid = int(self.mount_a.run_shell(['stat', '-c' '%g', subvolpath]).stdout.getvalue().strip())
+        self.assertEqual(uid, suid)
+        self.assertEqual(gid, sgid)
+
```
Comment (ajarr) [inline qa/tasks/cephfs/test_volumes.py]: In other tests we clean up the subvolumes.

**Comment by joscollin** [comment]
@ajarr Fixed

**Comment by ajarr** [comment]
Scheduled the run,
http://pulpito.front.sepia.ceph.com/rraja-2019-11-18_13:14:13-fs-rraja-subvolume-uid-gid-distro-basic-smithi/

**Comment by ajarr** [comment]
Tests have passed in the jobs.
@batrick ready to merge.

**Pull request merged**